File: quant1x/distributed/id/queue.cpp
// ID 队列的实现
#include <quant1x/distributed/id/queue.h>

#include <thread>

namespace quant1x::distributed::id {

Result<std::shared_ptr<Queue>> Queue::create(size_t capacity) {
    if (capacity == 0) {
        return Error::invalid_size();
    }
    // std::make_shared 无法访问私有构造, 显式 new 后立即交给 shared_ptr
    return std::shared_ptr<Queue>(new Queue(capacity));
}

Status Queue::try_push(const Id &value) {
    if (closed_.load(std::memory_order_acquire)) {
        return Error::closed();
    }
    // 底层 try_push 为 noexcept 实例化 (Id 为 trivial 类型), 满时返回 false
    if (inner_.try_push(value)) {
        return Error{};
    }
    return Error::queue_full();
}

Result<Id> Queue::try_pop() {
    Id value{};
    if (inner_.try_pop(value)) {
        return value;
    }
    // 底层不区分"空"与"已关闭且为空", 由关闭标志判定
    if (closed_.load(std::memory_order_acquire)) {
        return Error::closed();
    }
    return Error::queue_empty();
}

Status Queue::push(const Id &value) {
    while (true) {
        if (closed_.load(std::memory_order_acquire)) {
            return Error::closed();
        }
        if (inner_.try_push(value)) {
            return Error{};
        }
        // 队满, 让出后重试 (对齐 Go Push 的阻塞语义)
        std::this_thread::yield();
    }
}

Result<Id> Queue::pop() {
    while (true) {
        if (inner_.is_empty()) {
            if (closed_.load(std::memory_order_acquire)) {
                return Error::closed();
            }
            std::this_thread::yield();
            continue;
        }
        Id value{};
        if (inner_.try_pop(value)) {
            return value;
        }
        std::this_thread::yield();
    }
}

size_t Queue::len() const noexcept {
    return inner_.len();
}

size_t Queue::cap() const noexcept {
    return inner_.capacity();
}

void Queue::close() noexcept {
    closed_.store(true, std::memory_order_release);
    // 同时关闭底层队列, 使其内部不再接受写入语义
    inner_.close();
}

bool Queue::is_closed() const noexcept {
    return closed_.load(std::memory_order_acquire);
}

void Queue::wait_for_close() const {
    while (!inner_.is_empty()) {
        std::this_thread::yield();
    }
}

}  // namespace quant1x::distributed::id

File: quant1x/distributed/id/queue.rs
//! ID 队列 (对应 Go Queue, 底层复用 runtime::Queue 即 Vyukov 有界 MPMC 队列)
//!
//! 语义对齐 Go:
//! - `try_push` / `try_pop` 非阻塞;
//! - `push` / `pop` 阻塞 (忙等退避);
//! - `close` 后进入只读排空: 存量 ID 仍可消费, 队空后再返回 `Error::Closed`.

use std::sync::atomic::{AtomicBool, Ordering};

use crate::runtime;

use super::{Error, Id};

/// ID 队列
pub struct Queue {
    inner: runtime::Queue<Id>,
    closed: AtomicBool,
}

impl Queue {
    /// 创建队列, 容量向上取整到 2 的幂; 容量 0 返回错误
    pub fn new(capacity: usize) -> Result<Queue, Error> {
        if capacity == 0 {
            return Err(Error::InvalidSize);
        }
        Ok(Queue {
            inner: runtime::Queue::new(capacity),
            closed: AtomicBool::new(false),
        })
    }

    /// 非阻塞推入; 队列已满返回 `Error::QueueFull`, 已关闭返回 `Error::Closed`
    pub fn try_push(&self, value: Id) -> Result<(), Error> {
        if self.closed.load(Ordering::Acquire) {
            return Err(Error::Closed);
        }
        self.inner.try_push(value).map_err(|_| Error::QueueFull)
    }

    /// 非阻塞弹出; 队列为空返回 `Error::QueueEmpty`, 已关闭且空返回 `Error::Closed`
    pub fn try_pop(&self) -> Result<Id, Error> {
        match self.inner.try_pop() {
            Ok(value) => Ok(value),
            Err(()) => {
                if self.closed.load(Ordering::Acquire) {
                    Err(Error::Closed)
                } else {
                    Err(Error::QueueEmpty)
                }
            }
        }
    }

    /// 阻塞推入: 队列满时忙等腾位, 直到成功或队列关闭
    pub fn push(&self, value: Id) -> Result<(), Error> {
        loop {
            if self.closed.load(Ordering::Acquire) {
                return Err(Error::Closed);
            }
            if self.inner.push(value).is_ok() {
                return Ok(());
            }
            // 队满, 让出后重试 (对齐 Go Push 的阻塞语义)
            std::thread::yield_now();
        }
    }

    /// 阻塞弹出: 队列空时忙等数据, 直到成功; 已关闭且空返回 `Error::Closed`
    pub fn pop(&self) -> Result<Id, Error> {
        loop {
            if self.inner.is_empty() {
                if self.closed.load(Ordering::Acquire) {
                    return Err(Error::Closed);
                }
                std::thread::yield_now();
                continue;
            }
            if let Ok(value) = self.inner.pop() {
                return Ok(value);
            }
            std::thread::yield_now();
        }
    }

    /// 当前队列长度 (近似值)
    pub fn len(&self) -> usize {
        self.inner.len()
    }

    /// 队列容量 (2 的幂)
    pub fn cap(&self) -> usize {
        self.inner.cap()
    }

    /// 关闭队列: 进入只读排空模式
    pub fn close(&self) {
        self.closed.store(true, Ordering::Release);
    }

    /// 队列是否已关闭
    pub fn is_closed(&self) -> bool {
        self.closed.load(Ordering::Acquire)
    }

    /// 阻塞等待队列排空 (存量消费完毕)
    pub fn wait_for_close(&self) {
        while !self.inner.is_empty() {
            std::thread::yield_now();
        }
    }
}
